File: frontend/src/services/siteConfigService.ts
import api, { IS_STATIC } from './api';
import * as staticData from './staticDataService';

export interface FirstVisitModalConfig {
  enabled: boolean;
  title: string;
  content: string;
  min_stay_seconds: number;
  version: string;
}

export interface SiteComplianceConfig {
  enabled: boolean;
  icp_number: string;
  public_security_number: string;
}

export interface MaintenanceModeConfig {
  enabled: boolean;
  expected_end_time: string | null;
  version: string;
}

export interface TestEmailPayload {
  email: string;
}

interface ApiResponse<T> {
  success: boolean;
  data?: T;
  error?: {
    code: string;
    message: string;
  };
}

export const DEFAULT_FIRST_VISIT_MODAL_CONFIG: FirstVisitModalConfig = {
  enabled: false,
  title: '欢迎来到 HoYoMusic',
  content: '本站仅用于音乐欣赏与资料整理。请遵守相关法律法规。',
  min_stay_seconds: 5,
  version: '1',
};

export const DEFAULT_SITE_COMPLIANCE_CONFIG: SiteComplianceConfig = {
  enabled: false,
  icp_number: '',
  public_security_number: '',
};

export const DEFAULT_MAINTENANCE_MODE_CONFIG: MaintenanceModeConfig = {
  enabled: false,
  expected_end_time: null,
  version: '1',
};

export const siteConfigService = {
  async getPublicFirstVisitModal(): Promise<FirstVisitModalConfig> {
    if (IS_STATIC) {
      return staticData.getFirstVisitModalConfig();
    }

    const response = await api.get<ApiResponse<FirstVisitModalConfig>>('/public/site-config/first-visit-modal');
    if (response.data.success && response.data.data) {
      return response.data.data;
    }
    throw new Error(response.data.error?.message || 'Failed to fetch first-visit modal config');
  },

  async getAdminFirstVisitModal(): Promise<FirstVisitModalConfig> {
    const response = await api.get<ApiResponse<FirstVisitModalConfig>>('/settings/first-visit-modal');
    if (response.data.success && response.data.data) {
      return response.data.data;
    }
    throw new Error(response.data.error?.message || 'Failed to fetch first-visit modal config');
  },

  async updateAdminFirstVisitModal(payload: {
    enabled: boolean;
    title: string;
    content: string;
    min_stay_seconds: number;
  }): Promise<FirstVisitModalConfig> {
    const response = await api.put<ApiResponse<FirstVisitModalConfig>>('/settings/first-visit-modal', payload);
    if (response.data.success && response.data.data) {
      return response.data.data;
    }
    throw new Error(response.data.error?.message || 'Failed to update first-visit modal config');
  },

  async getPublicComplianceConfig(): Promise<SiteComplianceConfig> {
    if (IS_STATIC) {
      return staticData.getSiteComplianceConfig();
    }

    const response = await api.get<ApiResponse<SiteComplianceConfig>>('/public/site-config/compliance');
    if (response.data.success && response.data.data) {
      return response.data.data;
    }
    throw new Error(response.data.error?.message || 'Failed to fetch compliance config');
  },

  async getAdminComplianceConfig(): Promise<SiteComplianceConfig> {
    const response = await api.get<ApiResponse<SiteComplianceConfig>>('/settings/compliance');
    if (response.data.success && response.data.data) {
      return response.data.data;
    }
    throw new Error(response.data.error?.message || 'Failed to fetch compliance config');
  },

  async updateAdminComplianceConfig(payload: SiteComplianceConfig): Promise<SiteComplianceConfig> {
    const response = await api.put<ApiResponse<SiteComplianceConfig>>('/settings/compliance', payload);
    if (response.data.success && response.data.data) {
      return response.data.data;
    }
    throw new Error(response.data.error?.message || 'Failed to update compliance config');
  },

  async sendAdminTestEmail(payload: TestEmailPayload): Promise<{ message: string }> {
    const response = await api.post<ApiResponse<{ message: string }>>('/settings/test-email', payload);
    if (response.data.success && response.data.data) {
      return response.data.data;
    }
    throw new Error(response.data.error?.message || 'Failed to send test email');
  },

  async getPublicMaintenanceMode(): Promise<MaintenanceModeConfig> {
    if (IS_STATIC) {
      return staticData.getMaintenanceModeConfig();
    }

    const response = await api.get<ApiResponse<MaintenanceModeConfig>>('/public/site-config/maintenance');
    if (response.data.success && response.data.data) {
      return response.data.data;
    }
    throw new Error(response.data.error?.message || 'Failed to fetch maintenance mode config');
  },

  async getAdminMaintenanceMode(): Promise<MaintenanceModeConfig> {
    const response = await api.get<ApiResponse<MaintenanceModeConfig>>('/settings/maintenance');
    if (response.data.success && response.data.data) {
      return response.data.data;
    }
    throw new Error(response.data.error?.message || 'Failed to fetch maintenance mode config');
  },

  async updateAdminMaintenanceMode(payload: {
    enabled: boolean;
    expected_end_time: string | null;
  }): Promise<MaintenanceModeConfig> {
    const response = await api.put<ApiResponse<MaintenanceModeConfig>>('/settings/maintenance', payload);
    if (response.data.success && response.data.data) {
      return response.data.data;
    }
    throw new Error(response.data.error?.message || 'Failed to update maintenance mode config');
  },
};

File: client-desktop/src/HoYoMusic.Desktop.Infrastructure/Services/SiteConfigService.cs
using System.Net.Http.Json;
using System.Net.Http.Headers;
using System.Text.Json;
using HoYoMusic.Desktop.Core.Abstractions;
using HoYoMusic.Desktop.Core.Contracts;
using HoYoMusic.Desktop.Core.Models;

namespace HoYoMusic.Desktop.Infrastructure.Services;

public sealed class SiteConfigService : ISiteConfigService
{
    private readonly HttpClient _httpClient;
    private readonly ITokenStore _tokenStore;
    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web);

    public SiteConfigService(HttpClient httpClient, ITokenStore tokenStore)
    {
        _httpClient = httpClient;
        _tokenStore = tokenStore;
    }

    public Task<FirstVisitModalConfig> GetPublicFirstVisitModalAsync(CancellationToken cancellationToken = default)
    {
        return GetPublicConfigAsync<FirstVisitModalConfig>("public/site-config/first-visit-modal", "Failed to fetch first-visit config.", cancellationToken);
    }

    public Task<SiteComplianceConfig> GetPublicComplianceConfigAsync(CancellationToken cancellationToken = default)
    {
        return GetPublicConfigAsync<SiteComplianceConfig>("public/site-config/compliance", "Failed to fetch compliance config.", cancellationToken);
    }

    public Task<MaintenanceModeConfig> GetPublicMaintenanceModeAsync(CancellationToken cancellationToken = default)
    {
        return GetPublicConfigAsync<MaintenanceModeConfig>("public/site-config/maintenance", "Failed to fetch maintenance config.", cancellationToken);
    }

    public Task<FirstVisitModalConfig> GetAdminFirstVisitModalAsync(CancellationToken cancellationToken = default)
    {
        return GetAdminConfigAsync<FirstVisitModalConfig>("settings/first-visit-modal", "Failed to fetch first-visit config.", cancellationToken);
    }

    public Task<FirstVisitModalConfig> UpdateAdminFirstVisitModalAsync(FirstVisitModalConfig config, CancellationToken cancellationToken = default)
    {
        return PutAdminConfigAsync<FirstVisitModalConfig>("settings/first-visit-modal", new
        {
            enabled = config.Enabled,
            title = config.Title,
            content = config.Content,
            min_stay_seconds = config.MinStaySeconds,
        }, "Failed to update first-visit config.", cancellationToken);
    }

    public Task<SiteComplianceConfig> GetAdminComplianceConfigAsync(CancellationToken cancellationToken = default)
    {
        return GetAdminConfigAsync<SiteComplianceConfig>("settings/compliance", "Failed to fetch compliance config.", cancellationToken);
    }

    public Task<SiteComplianceConfig> UpdateAdminComplianceConfigAsync(SiteComplianceConfig config, CancellationToken cancellationToken = default)
    {
        return PutAdminConfigAsync<SiteComplianceConfig>("settings/compliance", new
        {
            enabled = config.Enabled,
            icp_number = config.IcpNumber,
            public_security_number = config.PublicSecurityNumber,
        }, "Failed to update compliance config.", cancellationToken);
    }

    public Task<MaintenanceModeConfig> GetAdminMaintenanceModeAsync(CancellationToken cancellationToken = default)
    {
        return GetAdminConfigAsync<MaintenanceModeConfig>("settings/maintenance", "Failed to fetch maintenance config.", cancellationToken);
    }

    public Task<MaintenanceModeConfig> UpdateAdminMaintenanceModeAsync(MaintenanceModeConfig config, CancellationToken cancellationToken = default)
    {
        return PutAdminConfigAsync<MaintenanceModeConfig>("settings/maintenance", new
        {
            enabled = config.Enabled,
            expected_end_time = string.IsNullOrWhiteSpace(config.ExpectedEndTime) ? null : config.ExpectedEndTime,
            message = config.Message,
        }, "Failed to update maintenance config.", cancellationToken);
    }

    public async Task<string> SendAdminTestEmailAsync(string email, CancellationToken cancellationToken = default)
    {
        using var request = await CreateAdminRequestAsync(HttpMethod.Post, "settings/test-email", cancellationToken);
        request.Content = JsonContent.Create(new { email }, options: JsonOptions);

        using var response = await _httpClient.SendAsync(request, cancellationToken);
        var envelope = await response.Content.ReadFromJsonAsync<ApiEnvelope<TestEmailResponseData>>(JsonOptions, cancellationToken);
        if (!response.IsSuccessStatusCode || envelope?.Success != true || envelope.Data is null)
        {
            throw await CreateApiExceptionAsync(response.StatusCode, envelope?.Error?.Message ?? "Failed to send test email.", envelope?.Error?.Code, cancellationToken);
        }

        return envelope.Data.Message;
    }

    private async Task<TConfig> GetPublicConfigAsync<TConfig>(string endpoint, string fallbackMessage, CancellationToken cancellationToken)
    {
        using var response = await _httpClient.GetAsync(endpoint, cancellationToken);
        var envelope = await response.Content.ReadFromJsonAsync<ApiEnvelope<TConfig>>(JsonOptions, cancellationToken);
        if (!response.IsSuccessStatusCode || envelope?.Success != true || envelope.Data is null)
        {
            throw new ApiException(envelope?.Error?.Message ?? fallbackMessage, envelope?.Error?.Code);
        }

        return envelope.Data;
    }

    private async Task<TConfig> GetAdminConfigAsync<TConfig>(string endpoint, string fallbackMessage, CancellationToken cancellationToken)
    {
        using var request = await CreateAdminRequestAsync(HttpMethod.Get, endpoint, cancellationToken);
        using var response = await _httpClient.SendAsync(request, cancellationToken);
        var envelope = await response.Content.ReadFromJsonAsync<ApiEnvelope<TConfig>>(JsonOptions, cancellationToken);
        if (!response.IsSuccessStatusCode || envelope?.Success != true || envelope.Data is null)
        {
            throw await CreateApiExceptionAsync(response.StatusCode, envelope?.Error?.Message ?? fallbackMessage, envelope?.Error?.Code, cancellationToken);
        }

        return envelope.Data;
    }

    private async Task<TConfig> PutAdminConfigAsync<TConfig>(string endpoint, object payload, string fallbackMessage, CancellationToken cancellationToken)
    {
        using var request = await CreateAdminRequestAsync(HttpMethod.Put, endpoint, cancellationToken);
        request.Content = JsonContent.Create(payload, options: JsonOptions);

        using var response = await _httpClient.SendAsync(request, cancellationToken);
        var envelope = await response.Content.ReadFromJsonAsync<ApiEnvelope<TConfig>>(JsonOptions, cancellationToken);
        if (!response.IsSuccessStatusCode || envelope?.Success != true || envelope.Data is null)
        {
            throw await CreateApiExceptionAsync(response.StatusCode, envelope?.Error?.Message ?? fallbackMessage, envelope?.Error?.Code, cancellationToken);
        }

        return envelope.Data;
    }

    private async Task<HttpRequestMessage> CreateAdminRequestAsync(HttpMethod method, string endpoint, CancellationToken cancellationToken)
    {
        var token = await _tokenStore.GetTokenAsync(cancellationToken);
        if (string.IsNullOrWhiteSpace(token))
        {
            throw new ApiException("Not authenticated.", "MISSING_TOKEN");
        }

        var request = new HttpRequestMessage(method, endpoint);
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
        return request;
    }

    private async Task<ApiException> CreateApiExceptionAsync(System.Net.HttpStatusCode statusCode, string message, string? code, CancellationToken cancellationToken)
    {
        if (statusCode == System.Net.HttpStatusCode.Unauthorized)
        {
            await _tokenStore.ClearTokenAsync(cancellationToken);
            return new ApiException("Session expired. Please login again.", "UNAUTHORIZED");
        }

        return new ApiException(message, code);
    }

    private sealed class TestEmailResponseData
    {
        public string Message { get; init; } = string.Empty;
    }
}

